### PR TITLE
Genetic algorithm to fine tune strategy parameters

### PR DESCRIPTION
**Pull request by arphen**
I don't think it's in a state to be merged, but i'd like to have a discussion

### CODE REVIEW

**Comment by XavierGeerinck** [comment]
Looking nice @arpheno! Do you got any results of it?

**Comment by arphen** [comment]
I have several tuned parameter constellations that for simulations are extremely profitable. One problem of simulations is that the slippage in real markets is much bigger than what is assumed in the simulator, so high frequency strategies really don't work out well.

I have 60 individuals across 60 days in three partitions running on my box right now, since yesterday night and waiting for results. These are trained across different currencies and different exchanges with a max-worst-case objective. 

Genetic algorithms are extremely good at learning crap, so if the data has quirks you get results over 12000% profit.

**Comment by XavierGeerinck** [comment]
Awesome! Would you be able to provide some documentation afterwards too? :)

**Comment by arphen** [comment]
#298

**Comment by DeviaVir** [comment]
This is great and impressive work, thank you for contributing! I personally think this is much better than the current backtesting script. This definitely needs some work (variable selectors and perhaps separate docker comes to mind) but seeing it work as-is right now is great.

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,97 @@
...
+
+RUN apt-get update
+RUN apt install -y  graphviz libgraphviz-dev pkg-config
+RUN PKG_CONFIG_ALLOW_SYSTEM_LIBS=OHYESPLEASE pip install pygraphviz
+ADD zen/requirements.txt /
+RUN pip3 install -r /requirements.txt
 WORKDIR /app
+RUN /usr/local/bin/npm install
+RUN npm update
+
```
Comment (DeviaVir) [inline Dockerfile:96]: I would suggest to just create your own docker.

Also, to run this, I had to run `mkdir -p logs/hof`, so perhaps a `RUN mkdir -p zen/logs/hof` here around the end?

File: fabfile.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import datetime
+import shlex
+
+import subprocess
+from fabric.api import run, cd,local
+
+
+
```
Comment (DeviaVir) [inline fabfile.py:8]: nit: one newline too much

File: zen/.idea/inspectionProfiles/profiles_settings.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<component name="InspectionProjectProfileManager">
+  <settings>
+    <option name="useProjectProfile" value="false" />
+    <option name="USE_PROJECT_PROFILE" value="false" />
+    <version value="1.0" />
+  </settings>
+</component>
```
Comment (DeviaVir) [inline zen/.idea/inspectionProfiles/profiles_settings.xml:7]: this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

File: zen/.idea/misc.xml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 2.7.12 (/usr/bin/python2.7)" project-jdk-type="Python SDK" />
+</project>
```
Comment (DeviaVir) [inline zen/.idea/misc.xml:4]: this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

File: zen/.idea/modules.xml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectModuleManager">
+    <modules>
+      <module fileurl="file://$PROJECT_DIR$/.idea/zen.iml" filepath="$PROJECT_DIR$/.idea/zen.iml" />
+    </modules>
+  </component>
+</project>
```
Comment (DeviaVir) [inline zen/.idea/modules.xml:8]: this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

File: zen/.idea/zen.iml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+  <component name="PackageRequirementsSettings">
+    <option name="requirementsPath" value="" />
+  </component>
+  <component name="TestRunnerService">
+    <option name="projectConfiguration" value="py.test" />
+    <option name="PROJECT_TEST_RUNNER" value="py.test" />
+  </component>
+</module>
```
Comment (DeviaVir) [inline zen/.idea/zen.iml:15]: this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

File: zen/.idea/workspace.xml [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+        <state relative-caret-position="239">
+          <caret line="19" column="0" lean-forward="false" selection-start-line="19" selection-start-column="0" selection-end-line="19" selection-end-column="0" />
+          <folding>
+            <element signature="e#0#15#0" expanded="true" />
+          </folding>
+        </state>
+      </provider>
+    </entry>
+  </component>
+</project>
```
Comment (DeviaVir) [inline zen/.idea/workspace.xml:736]: this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

File: zen/conf.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import random
+
+
+
+
```
Comment (DeviaVir) [inline zen/conf.py:5]: nit: too many newlines

File: zen/evaluation.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            raise ValueError(colored(f"I don't understand {param} please add it to evaluation.py", 'red'))
+        return param, res
+
+
+def fuzz_product(product: Product) -> List[Selector]:
+    """ We want to run against multiple selectors for a product to reduce sampling error.
+    >>>fuzz_product('USD')"""
+    selectors = {
+        'BTC-CUR': ['gdax.BTC-USD', 'gdax.BTC-EUR', 'gdax.BTC-GBP'],
+        'ETH-BTC': ['poloniex.ETH-BTC', 'gdax.ETH-BTC']
```
Comment (DeviaVir) [inline zen/evaluation.py]: I understand why this is a constant now (for demo purposes) but if we plan on merging this in at some point, we should try and use the `list-selectors` from zenbot, perhaps?

File: fabfile.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        print(total)
+        run(total)
+
+def backfill_remote(TOTAL_DAYS):
+    products = ['gdax.BTC-EUR','gdax.BTC-USD','gdax.BTC-GBP']+['gdax.ETH-BTC','poloniex.ETH-BTC']
+    for instrument in products:
+        cmd = '/app/zenbot.sh backfill {instrument} --days {days}'.format(days=TOTAL_DAYS, instrument=instrument)
+        remote(cmd,'backfill_'+instrument)
+def backfill_local(TOTAL_DAYS):
+    products = ['gdax.BTC-EUR','gdax.BTC-USD','gdax.BTC-GBP']+['gdax.ETH-BTC','poloniex.ETH-BTC']
```
Comment (DeviaVir) [inline fabfile.py:33]: I understand why this is a constant now (for demo purposes) but if we plan on merging this in at some point, we should try and use the `list-selectors` from zenbot, perhaps?

**Comment by nao0811ta** [comment]
question.

It have never been evaluated during few sleeping hours.
```
[ec2-user@ip-10-166-11-188 zen]$ python -m scoop main.py ETH-BTC 14 10 macd
[2017-06-20 15:51:58,504] launcher  INFO    SCOOP 0.7 1.1 on linux using Python 3.6.1 (default, Jun 20 2017, 15:15:55) [GCC 4.8.3 20140911 (Red Hat 4.8.3
-9)], API: 1013
[2017-06-20 15:51:58,504] launcher  INFO    Deploying 2 worker(s) over 1 host(s).
[2017-06-20 15:51:58,505] launcher  INFO    Worker distribution:
[2017-06-20 15:51:58,505] launcher  INFO       127.0.0.1:       1 + origin
MAKE SURE YOU RUN fab backfill_local:<days>
otherwise it's all crap
Starting evolution....
Evaluating 10 individuals over 14 days in 2 partitions.
Mating function is <function cxTwoPoint at 0x7fc3fea96268>
Mutating function is functools.partial(<function mutGaussian at 0x7fc3fbee88c8>, mu=0, sigma=20, indpb=0.7)
Generating an initial valid population, this may take a while...
Summer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxx-xxxxxSummer is here, need to evaluate 10 individuals
<abbr asame output>
```

**Comment by nao0811ta** [comment]
The reason why is the trades per day was few.

**Comment by nao0811ta** [comment]
FileNotFoundError: [Errno 2] No such file or directory: '/app/zenbot.sh'

I wish that zenbot.sh is relative path
and saimulation saving path also([#317](https://github.com/carlos8f/zenbot/issues/317)).

**Comment by arphen** [comment]
I screwed up something in master, my bad. I'm fixing it right now.

**Comment by nao0811ta** [comment]
It takes so long time. So 
How about save and restart function?

**Comment by nao0811ta** [comment]
```
  File "main.py", line 41, in <module>
    res = main(INSTRUMENT, TOTAL_DAYS, popsize, strategy)
  File "main.py", line 24, in main
    res = evolve(evaluate, Individual, popsize)
  File "/home/ec2-user/workspace/zen4-polo/zen/evolution/__init__.py", line 27, in evolve
    return algorithm(toolbox, stats, history, hof,popsize)
  File "/home/ec2-user/workspace/zen4-polo/zen/evolution/core.py", line 16, in algorithm
    evaluate(would_be, toolbox)
  File "/home/ec2-user/workspace/zen4-polo/zen/evolution/core.py", line 40, in evaluate
    for ind, fit in zip(invalid_ind, fitnesses):
  File "/home/ec2-user/.pyenv/versions/zen-polo/lib/python3.6/site-packages/scoop/futures.py", line 102, in _mapGenerator
    for future in _waitAll(*futures):
  File "/home/ec2-user/.pyenv/versions/zen-polo/lib/python3.6/site-packages/scoop/futures.py", line 358, in _waitAll
    for f in _waitAny(future):
  File "/home/ec2-user/.pyenv/versions/zen-polo/lib/python3.6/site-packages/scoop/futures.py", line 335, in _waitAny
    raise childFuture.exceptionValue
IndexError: Cannot choose from an empty sequence
```

**Comment by arphen** [comment]
Yep sorry guys for the inconvenience. I fucked up the master branch and learned yet again not to commit to master directly. Multiprocessing is more exciting than it should be.

Things are working again.

**Comment by nao0811ta** [comment]
@arpheno 
thanks!! it works again.

**Comment by arphen** [comment]
Here's some results:

```
user@genericbox:~/zenbot$ tail -f ETH*30*.out
==> ETH-BTC_macd_50_30_2017-06-22-18-04.out <==
...

Generation 156 {'avg': -4.3447320162938903, 'std': 23.51224732056831, 'min': -39.009999999999998, 'max': 19.99943054239105, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
8 individuals have mutated.
Summer is here, evaluating our new arrivals...
                  |
........x........
==> ETH-BTC_rsi_50_30_2017-06-22-18-06.out <==
...

Generation 191 {'avg': -13.496419041828347, 'std': 13.552027774607932, 'min': -31.670000000000002, 'max': 1.0922793968148585, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
9 individuals have mutated.
Summer is here, evaluating our new arrivals...
                   |
................
==> ETH-BTC_sar_50_30_2017-06-22-13-42.out <==
...

Generation 366 {'avg': -3.786960662691008, 'std': 9.7398293425581581, 'min': -30.240098048429321, 'max': 9.0500729772017543, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
11 individuals have mutated.
Summer is here, evaluating our new arrivals...
                     |
...................
==> ETH-BTC_trend_ema_50_30_2017-06-22-18-06.out <==
...

Generation 162 {'avg': -15.808718271857389, 'std': 21.800960754046944, 'min': -100.69314718055995, 'max': 4.2826980234809708, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
9 individuals have mutated.
Summer is here, evaluating our new arrivals...
                   |

```

**Review by julymetodiev — APPROVED** [review]

**Comment by nao0811ta** [comment]
Thank you all
DQN Awesome..

Do you know why happen like below?
```
python -m scoop main.py ETH-BTC 10 30 macd backfill
[2017-06-28 15:51:12,306] launcher  INFO    SCOOP 0.7 1.1 on linux using Python 3.6.1 (default, Jun 20 2017, 15:15:55) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)], API: 1013
[2017-06-28 15:51:12,306] launcher  INFO    Deploying 2 worker(s) over 1 host(s).
[2017-06-28 15:51:12,306] launcher  INFO    Worker distribution:
[2017-06-28 15:51:12,306] launcher  INFO       127.0.0.1:       1 + origin 
Auto backfilling is ON
Executing inital backfilling, please wait...
Traceback (most recent call last):
  File "/home/ec2-user/.pyenv/versions/3.6.1/envs/zen-polo/lib/python3.6/site-packages/fabric/main.py", line 672, in main
    docstring, callables, default = load_fabfile(fabfile)
  File "/home/ec2-user/.pyenv/versions/3.6.1/envs/zen-polo/lib/python3.6/site-packages/fabric/main.py", line 173, in load_fabfile
    imported = importer(os.path.splitext(fabfile)[0])
  File "/home/ec2-user/workspace/zen4-polo/fabfile.py", line 5, in <module>
    from zen import path, pyexc, selectors
  File "/home/ec2-user/workspace/zen4-polo/zen/__init__.py", line 1, in <module>
    from conf import path, pyexc, selectors
ImportError: cannot import name 'path'
```

**Comment by Yroethiel** [comment]
Any suggestions why "docker-compose build" fails with this Dockerfile?

It breaks on Step 4/17:

`RUN apt-get update && apt-get install -y --no-install-recommends tcl tk && rm -rf /var/lib/apt/lists/*`

The error I get is:

`ERROR: Service 'server' failed to build: the command '/bin/sh apt-get update && apt-get install -y --no-install-recommends tcl tk && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 1 `

Latest version of Debian Jessie, latest python, latest npm, latest Docker. Tried on ubuntu as well.

**Comment by arphen** [comment]
It really shouldn't happen. Did you git pull the newest version? Somethings
not right with the last PR

**Comment by Yroethiel** [comment]
Yeah, has me confused as well. All cloned via git, using today's release. Same issue on the release from May 22.

**Comment by sandroid10** [comment]
@arpheno  i am getting this error inside docker ( i did backfill) 

Evaluating 10 individuals over 3 days in 2 partitions.
Mating function is <function cxTwoPoint at 0x7f9dc17ce400>
Mutating function is functools.partial(<function mutGaussian at 0x7f9dc01efae8>, mu=0, sigma=20, indpb=0.3)
Sampling an initial valid population, this may take a while...
Currently 0 valid individuals
          |
...x......Currently 5 valid individuals
          |
..........Traceback (most recent call last):
  File "/usr/local/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 302, in <module>
    b.main()
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 92, in main
    self.run()
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 290, in run
    futures_startup()
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 271, in futures_startup
    run_name="__main__"
  File "/usr/local/lib/python3.6/site-packages/scoop/futures.py", line 64, in _startup
    result = _controller.switch(rootFuture, *args, **kargs)
  File "/usr/local/lib/python3.6/site-packages/scoop/_control.py", line 253, in runController
    raise future.exceptionValue
  File "/usr/local/lib/python3.6/site-packages/scoop/_control.py", line 127, in runFuture
    future.resultValue = future.callable(*future.args, **future.kargs)
  File "/usr/local/lib/python3.6/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/usr/local/lib/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/local/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "main.py", line 42, in <module>
    res = main(INSTRUMENT, TOTAL_DAYS, popsize, strategy)
  File "main.py", line 25, in main
    res = evolve(evaluate, Andividual, popsize)
  File "/app/zen/evolution/__init__.py", line 18, in evolve
    return algorithm(cls, popsize, futures.map, evaluate, select, breed, mutate, stats, history, hof)
  File "/app/zen/evolution/core.py", line 21, in algorithm
    log_stuff(g, history, hof, population, stats)
  File "/app/zen/evolution/utils.py", line 30, in log_stuff
    hof.persist()
  File "/app/zen/halloffame.py", line 27, in persist
    with open('logs/hof/{runid}.txt'.format(runid=runid), 'w') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'logs/hof/3520.txt'
[2017-07-25 17:02:54,946] launcher  (127.0.0.1:37786) INFO    Root process is done.
[2017-07-25 17:02:54,946] launcher  (127.0.0.1:37786) INFO    Finished cleaning spawned subprocesses.

**Comment by Ekliptor** [comment]
What's the difference between elites and specialists in your selection algorithm? I understand the goal of the algorithm is to find the most profitable strategy parameters. So what is the criteria for "profitable" for elites and specialists?

**Comment by sandroid1** [comment]
@Ekliptor i just cloned this repo https://github.com/arpheno/zenbot , setup conf.js and run backfill inside docker

**Comment by DeviaVir** [comment]
@arpheno I would like to merge your PR, but there are still a few unhandled comments and things I would like to see changed.

The most important one, in my eyes, is that this should live in the `/scripts/` folder and not touch so many other files that have relatively little to do with the proposed additions.

**Review by DeviaVir — CHANGES_REQUESTED** [review]
@arpheno I would like to merge your PR, but there are still a few unhandled comments and things I would like to see changed.

The most important one, in my eyes, is that this should live in the `/scripts/` folder and not touch so many other files that have relatively little to do with the proposed additions.

File: Dockerfile (hunk 1 above)
Comment (ivanmladek) [reply Dockerfile:96]: I get a weird error related to npm in docker. When you guys were building your docker, did you stumble upon this? 
server_1        | module.js:471
server_1        |     throw err;
server_1        |     ^
server_1        | 
server_1        | Error: Cannot find module 'semver'

Comment (nodevar) [reply Dockerfile:96]: If you are using arphenos dockerfile you need to copy at least package.json into you workdir before the npm install command.

or just copy the app into the working dir: 
COPY . /app

**Comment by evseevnn-zz** [comment]
@arpheno That work? how soon you wanna solve problem of PR?
May be somebody can help?

**Comment by DeviaVir** [comment]
Superseded by https://github.com/carlos8f/zenbot/pull/598

**Pull request closed without merging**